### DISCUSSION TITLE
How to introduce external dependencies as part of the handler definition?

### DISCUSSION THREAD

**Discussion by theogravity** [daltonmenezes/interprocess#14] [Q&A]
Let's say that I have a handler that requires a `browserWindow` to be fed into it. Given the way the handlers are defined via `createInterprocess`, there's no way to introduce external dependencies like this as part of the initial handler definition.

I've found that I have to leave it as a dummy implementation then wrap the handler as part of a closure to do that, ex:

```typescript
export const { ipcMain } = createInterprocess({
  main: {
    askForPermission: async (_, _permissionType: MediaPermissionsType): Promise<MediaPermissionsStatus> => {
     // return a dummy result so typescript doesn't complain
      return MediaPermissionsStatus.Granted;
    },
    /**
```

```typescript
export function handleAskForPermission(browserWindow: BrowserWIndow) {
  IpcMain.handle.askForPermission(async (_, { data: permissionTypes }) => {
      // now I can work with the browser window here
  });
}
```

Is there a better way to do it?

**Comment by daltonmenezes** [comment]
Hi @theogravity , currently in one project I'm working in, I'm avoiding dummy implementation in favor to do something like this:

`src/shared/ipcs/index.ts`
```ts
import { createInterprocess } from 'interprocess'

import { openDialog } from 'main/modules/dialog/open'

export const { ipcMain } =
  createInterprocess({
    main: {
      openDialog,
    },
  })
```

`main/modules/dialog/open.ts`
```ts
import { OpenDialogOptions, dialog } from 'electron'

export async function openDialog(
  _: Electron.IpcMainInvokeEvent,
  data: OpenDialogOptions
) {
  return dialog.showOpenDialog(data)
}
```
`main/modules/dialog/index.ts`
```ts
import { ipcMain } from 'shared/ipcs'

export function registerDialogModule() {
  ipcMain.handle.openDialog()
}
```
When I need to handle more things before returning the response to the renderer process, as mentioned in the docs, we can handle like this:
```ts
export function registerGetPingModule(window: Electron.BrowserWindow) {
  ipcMain.handle.getPing(async (_, { getPing, data }) => {
    // call the registered handler if needed
    const response = await getPing(_, data)

    await ipcMain.invoke.getPong(mainWindow, 'pong')

    return response
  })
}
```

**Reply by theogravity** [reply]
Thanks for the suggestion, but it unfortunately feels like an extra step or two to do, or rather there's just a lot of indirection going on that makes it hard to follow since there's four areas to work with.
